experimental/InvariantTheory/invariant_rings.jl

```diff
@@ -423,3 +359,107 @@ function heisenberg_group(n::Int)
+  R = polynomial_ring(IR)
+  K = coefficient_ring(IR)
+  # Need an extra coercion here while waiting for https://github.com/Nemocas/AbstractAlgebra.jl/pull/1009
+  #return to_univariate(S, R(mol[1, 1]))//to_univariate(S, R(mol[1, 2]))
+  return S(to_univariate(S, R(mol[1, 1])))//S(to_univariate(S, R(mol[1, 2])))
+end
+
+function _molien_series_via_singular(S::PolyRing, IR::InvRing{T}) where {T <: Union{Nemo.GaloisField, Nemo.GaloisFmpzField}}
+  @assert !ismodular(IR)
+  mol = __molien_series_via_singular(IR)
+
+  # TODO: Write a conversion from Singular number fields to Nemo ones
+  Qx, x = PolynomialRing(FlintQQ, "x", cached = false)
+  K, a = number_field(Qx(Singular.n_transExt_to_spoly(Singular.modulus(coefficient_ring(parent(mol[1, 1]))))), "a", cached = false)
+  R, y = PolynomialRing(K, ["y"], cached = false)
+
+  # Need an extra coercion here while waiting for https://github.com/Nemocas/AbstractAlgebra.jl/pull/1009
+  #return to_univariate(S, R(mol[1, 1]))//to_univariate(S, R(mol[1, 2]))
+  return S(to_univariate(S, R(mol[1, 1])))//S(to_univariate(S, R(mol[1, 2])))
+end
+
+function _molien_series_char0(S::PolyRing, I::InvRing)
+  G = group(I)
+  n = degree(G)
+  Gp, GtoGp = isomorphic_group_over_finite_field(G)
+  K = coefficient_ring(I)
+  C = conjugacy_classes(Gp)
+  res = zero(FractionField(S))
+  for c in C
+    g = GtoGp\(representative(c))
+    f = charpoly(S, g.elm)
+    res = res + length(c) * 1//reverse(f)
+  end
+  return divexact(res, order(Gp))
+end
+
+function _molien_series_charp_nonmodular_via_gap(S::PolyRing, I::InvRing)
+  G = group(I)
+  @assert G isa MatrixGroup
+  t = GAP.Globals.CharacterTable(G.X)
+  chi = [GAP.Globals.BrauerCharacterValue(GAP.Globals.Representative(c))
+         for c in GAP.Globals.ConjugacyClasses(t)]
+  info = GAP.Globals.MolienSeriesInfo(GAP.Globals.MolienSeries(t,
+                                                               GAP.GapObj(chi)))
+  num = S(Vector{fmpz}(GAP.Globals.CoefficientsOfUnivariatePolynomial(info.numer)))
+  den = S(Vector{fmpz}(GAP.Globals.CoefficientsOfUnivariatePolynomial(info.denom)))
+  return num//den
+end
+
+function molien_series(S::PolyRing, I::InvRing)
+  if characteristic(coefficient_ring(I)) == 0
+    return _molien_series_char0(S, I)
+  else
+    if !ismodular(I)
+      return _molien_series_charp_nonmodular_via_gap(S, I)
+    else
+      throw(NotImplemented())
+    end
+  end
+end
+
+doc"""
+    molien_series([S::PolyRing], I::InvRing)
+
+Return the Molien series of `I` as a rational function. The invariant ring must
+be non-modular.
+
+# Examples
+```jldoctest
+julia> K, a = CyclotomicField(3, "a");
+
+julia> M1 = matrix(K, [0 0 1; 1 0 0; 0 1 0]);
+
+julia> M2 = matrix(K, [1 0 0; 0 a 0; 0 0 -a-1]);
+
+julia> G = MatrixGroup(3, K, [M1, M2]);
+
+julia> IR = invariant_ring(G);
+
+julia> molien_series(G)
+(-t^6 + t^3 - 1)//(t^9 - 3*t^6 + 3*t^3 - 1)
+```
+"""
+function molien_series(I::InvRing)
+  S, t = PolynomialRing(coefficient_ring(I), "t", cached = false)
```

Comment:
I'm confused. Doesn't one have to use a different (characteristic 0) coefficient ring for the Hilbert series in the characteristic p case?

Reply:
Indeed, good catch!

Also, if we want to keep a factorized representation, the coefficient ring must be some a priori unknown number field (of course as a rational function, expanded, it is defined over Q). I guess all of these aspects need to be taken care of by the to-be-added HilbertSeries type.

Reply:
Hm, oops, of course! I guess we should always use a polynomial ring over the rationals by default? If I understand it correctly, the Molien series will always be a rational function over the rationals?

Reply:
I think the idea will be to compute the factorization from the rational function f/g with f, g in Q[t]. @tthsqe12 is looking into it. In the meantime, I will change it to always use a rational polynomial ring by default.

Reply:
Actually, for most uses, I guess it's fine to do all over Q: If I am not mistaken, the Hilbert series always can be written as `f(t) / [ (1-t^{d_1} ) \cdots (1-t^{d_n}) ]` for suitable exponents d_1, ..., d_n. While one could factor this further over a suitable cylotomic field extension, I don't think this is necessary. Indeed, for many applications, just getting `f` and the tuple `(d_1, ..., d_n)` will be enough.

Reply:
@fingolfin Is there some canonical representation of the Molien series? I would like to at least be able to tell the order of growth of the coefficients just by looking at the denominator? i.e. 1/(1-x) -> bounded, 1/((1-x)*(1-x^2)) -> degree 1, ect

Reply:
No, in general there is not. That's why this is all some magic way to print it nicely, but without mathematical justification.

Reply:
Here is another example
```
  1-x+x^2           1+x^3
------------ = --------------
(1-x)(1-x^3)   (1-x^2)(1-x^3)
```

Probably we would go with the first one, but the second one allows for an easy proof that the coefficients are non-negative.

Reply:
Regarding the coefficient ring: Singular constructs a cyclotomic field in this situation, but I have the impression that for GAP QQ is enough.

Reply:
@thofma so you have absolutely no desire to see positive coefficients on top?

Reply:
Maybe we should ask someone who would like to have those. @fingolfin Any wishes on how it should look like? Or maybe @wdecker?

Reply:
Let me recall: In the 19th century, people played with different representations of the Hilbert series in order to make guesses about where to find generators for the invariant ring. In the context of computing invariants of finite groups, the only reasonable factorization is that corresponding to the degrees of the primary invariants. So how about leaving the Molien series unfactored until the primary invariants are computed, and cache the meaningful factorization then.

Reply:
Thanks for the clarification. I was getting anxious over all of the different possible representations.

Reply:
@tthsqe12: If unfactored, the 1 in the denominator should be 1 and not -1. This fits with the eventual meaningful factorization which exists over ZZ.

Reply:
Sorry, I was to quick. Of course, there is the theorem of Hilbert-Serre and thus another meaningful way of representing the Hilbert series which reflects the degrees of the fundamental invariants in the denominator (which do not(!!!) depend on the choice of fundamental invariants). Often, however, we do not wish to compute the fundamental invariants. But if they are computed, we could  also cache this factorization. I will think about this more carefully.

Reply:
Sorry, I was to quick. Of course, there is the theorem of Hilbert-Serre and thus another meaningful way of representing the Hilbert series which reflects the degrees of the fundamental invariants in the denominator (which do not(!!!) depend on the choice of fundamental invariants). Often, however, we do not wish to compute the fundamental invariants. But if they are computed, we could  also cache this factorization. I will think about this more carefully.

Reply:
Sorry, I was to quick. Of course, there is the theorem of Hilbert-Serre and thus another meaningful way of representing the Hilbert series which reflects the degrees of the fundamental invariants in the denominator (which do not(!!!) depend on the choice of fundamental invariants). Often, however, we do not wish to compute the fundamental invariants. But if they are computed, we could  also cache this factorization. I will think about this more carefully.